### PR TITLE
adds b3 specific formatting for trace_id

### PR DESCRIPTION
**Pull request by castengo**
When reported to signalFX, trace_ids padded with 0 are dropped. This PR makes it so the propagation format can be configured, defaulting to tracecontext so it doesn't break current implementations.

I am not an erlang dev so any lang suggestions are welcome.

### CODE REVIEW

File: src/oc_reporter_zipkin.erl [hunk 1]
```diff
@@ -78,6 +80,15 @@ zipkin_span(Span, LocalEndpoint) ->
        <<"tags">> => to_tags(Span#span.attributes) %% TODO: merge with oc_tags?
      }.
 
+format_trace_id(TraceId, tracecontext) ->
+    iolist_to_binary(io_lib:format("~32.16.0b", [TraceId]));
+
+format_trace_id(TraceId, b3) ->
+    iolist_to_binary(io_lib:format("~.16.0b", [TraceId]));
```
Comment (castengo) [inline src/oc_reporter_zipkin.erl]: I wasn't sure how to use the format lib, but not specifying the width made it so if it was 16 chars it wouldn't get padded with 0s.

**Comment by tsloughter** [comment]
Hm, interesting. The B3 spec doesn't seem say to leave out the 0 padding?

Are your traces started by opencensus? Just making sure that the trace ids are 128 bit. If the trace is started another way and ends up as 64 bit trace id I could see SignalFX expecting 16 char hexadecimal trace id instead of 32 and this fixing it but if that isn't the case I'm surprised SignalFX cares.

**Comment by castengo** [comment]
Yeah so the traces are being generated in other services, not sure if through opencensus or not, but we receive them in b3 format which could be 64 or 128 bit according to the spec. I am not completely sure SignalFX drops the traces padded with 0s but they definitely don't match with the reported trace ids from our other services bc they are being reported without the 0 padding. 

I usually write in Elixir, this is actually my first erlang PR ever so if there is a better way to encode the trace_id rather  than this `"~.16.0b"` that would be great.

**Comment by tsloughter** [comment]
If the issue is that the trace id's are 64 bit instead of 128 then I'd suggest changing the new format to use `"~16.16.0b"`.

I'd also get rid of the configuration stuff and just have it check if the trace id is 128 or 64 bit and choose the format based on that.

Using a 64 bit trace id isn't advised though, it could end up breaking else where when using opencensus because opencensus assumes 128 bit.

But if this PR is changed to just use the format based on the size of the id I think I can accept this even if opencensus technically says it should be 128 bit, since we are focued on https://opentelemetry.io/ I don't find it as important to keep this lib strict.

**Comment by castengo** [comment]
Do you think anyone using the lib would be relying on that 64 bit trace_id being formatted to a 128 bit? That's why I thought about the configuration, bc I was nervous to break any existing implementations.

If you don't think that's a concern, then I can go ahead with the changes.

Thank you!!

**Comment by tsloughter** [comment]
I don't think there should be an issue since this is only exporting to zipkin and zipkin should be fine with it.

Changes look good, I'll merge this today.

**Comment by castengo** [comment]
I'm going to close this MR, bc it doesn't conform to the opencensus specs.

**Comment by tsloughter** [comment]
Oh, do you not need this anymore?

**Comment by castengo** [comment]
Yes still need it but I'm just not sure about going against the spec 😅 But if you think theres no issues, lets do it!

**Comment by tsloughter** [comment]
Yea, I think it is fine since it is just the exporter going to zipkin. You might have other issues somewhere in the system because OC expects/requires 128 bit, so best to replace whatever is currently producing 64 bit trace ids, but I think this patch is fine if it provides you a solution in the mean time.

**Review by tsloughter — APPROVED** [review]

**Comment by castengo** [comment]
thank you 🙇‍♀️! Our other apps are using OpenTracing which doesn't seem to be strict about the length. But hopefully soon we all start using OpenTelemetry

**Pull request merged**